lib/src/conflicts.rs

```diff
@@ -454,25 +483,46 @@ pub fn update_conflict_from_content(
         // with conflict markers. Update the Conflict object with the new
         // FileIds.
         for (i, buf) in removed_content.iter().enumerate() {
-            let file_id = store.write_file(path, &mut buf.as_slice())?;
-            if let TreeValue::File { id, executable: _ } = &mut conflict.removes[i].value {
-                *id = file_id;
-            } else {
-                // TODO: This can actually happen. We should check earlier
-                // that the we only attempt to parse the conflicts if it's a
-                // file-only conflict.
-                panic!("Found conflict markers in merge of non-files");
+            match &conflict.removes()[i] {
+                Some(TreeValue::File { id: _, executable }) => {
+                    let file_id = store.write_file(path, &mut buf.as_slice())?;
+                    let new_value = TreeValue::File {
+                        id: file_id,
+                        executable: *executable,
+                    };
+                    conflict.set_remove(i, Some(new_value));
+                }
+                None if buf.is_empty() => {
+                    // The missing side of a conflict is still represented by
+                    // the empty string we materialized it as => nothing to do
+                }
+                _ => {
+                    // The user edited a non-file side. We consider the conflict resolved.
```

Comment:
If I understand what's going on correctly, I would have this comment be much more forceful in tone:

"This is an exceptional situation; we treat it as though the conflict markers didn't have the correct arity"

I think the correct thing to do here would be to return some kind of an error value and warn the user that something went wrong reading this conflict. Perhaps we should leave a TODO here, and in the two other places we `return Ok(None)`, that this should become an error?

Reply:
Perhaps we could also mention that this is an exceptional situation because we can't tell if the file on that side should be executable or not. Otherwise, we'd just use the text the user gave us.

Reply:
Your comments don't quite cover the case I had in mind; I didn't explain it very well.

I was thinking of a case when you get some text instead of no text for the deleted file. So, instead of the string you have in the test:

```
b"<<<<<<<\n%%%%%%%\n line 1\n-line 2 before\n+line 2 modified after\n line 3\n+++++++\n>>>>>>>\n",
```

you get something like:

```
b"<<<<<<<\n%%%%%%%\n line 1\n-line 2 before\n+line 2 modified before\n line 3\n+++++++\nSome text instead of the deleted file\n>>>>>>>\n",
```

(assuming I'm using the markers correctly).

Perhaps this is too insane to consider?

Reply:
This is, in fact, extremely similar to the case where the conflict being read has too many sides, so it seems completely appropriate to do the same thing here as in that case (which is exactly what happens now). However, I still think that, in the ideal world, this situation would result in some sort of error/warning as opposed to quietly considering the conflict to be resolved with the crazy conflict markers remaining.

Reply:
Ah, that's the `None if !buf.is_empty()`, which currently matches the wildcard pattern, and which will remain after the TODO I mentioned. Yes, I think that's similar to the wrong-arity case.

Btw, we talked about conflicts in meeting at Google a few days ago and some people remarked that it might be confusing that we consider broken conflict markers as resolved. I responded that I didn't know what else we could do with such conflicts. But last night I realized that it might be a good idea to continue to consider the file a conflict as long as we find anything that even remotely looks like a conflict marker in the file (if it has `<<<<<<<`/ `+++++++`. `%%%%%%%`, or `>>>>>>>` in it). Even if we can't parse the conflict markers, we can just say that all sides had the same contents. I'll file a bug about that.

Reply:
Never mind, that would probably just be more confusing :( It would mean that a broken
```
<<<<<<<
+++++++
left
>>>>>>>
```
would be replaced by
```
<<<<<<<
+++++++
left
%%%%%%%
-left
+left
>>>>>>>
```
next time we materialize it (e.g. when you run `jj diff`).

Reply:
Hmm, that's actually not true. Since we don't save conflicts at the hunk-level, we could just save the whole broken conflict marker as regular content and it would be materialized as a broken conflict marker. We could still say that the file has a conflict even if all sides are the same. However, with #1624, we lose that option 😞 We could add a commit-level list of files with conflicts but that seems uglier. I'm not sure if there's a better way.

Reply:
> Btw, we talked about conflicts in meeting at Google a few days ago and some people remarked that it might be confusing that we consider broken conflict markers as resolved. I responded that I didn't know what else we could do with such conflicts. 

Yes, that's my concern as well. I think we should report an error if at all possible. The actual behavior of keeping the conflict markers and resolving the file seems OK. One way of warning the user, though, might be to add a header to it, something along the lines of

```
JJ: jj failed to read the conflict markers in this file. You will need to
JJ: either `jj restore` this file and resolve the conflict again, or you
JJ: can edit this file to its desired end state.
```

Reply:
I'll probably attempt to make this happen sometime.

Reply:
Alternatively, we might be able to have `TreeState::snapshot()` report any files it failed to parse conflict markers in. Your proposal has the advantage that the user can't miss it.